bin/configure-tower-projects.py

```diff
@@ -0,0 +1,618 @@
+
+
+class Projects:
+    def __init__(self, config_directory: str) -> None:
+        """Create Projects instance
+
+        Args:
+            config_directory (str): Directory containing project config files
+        """
+        self.config_directory = config_directory
+        self.users_per_project = self.extract_users()
+
+    def list_projects(self) -> Iterator[str]:
+        """List all project YAML configuration files
+
+        Yields:
+            Iterator[str]:
+                Each element is a YAML filepath as a str
+        """
+        # Obtain a list of config files from the given directory
+        for dirpath, _, filenames in os.walk(self.config_directory):
+            for filename in filenames:
+                filepath = os.path.join(dirpath, filename)
+                if filename.endswith("-project.yaml"):
+                    yield filepath
+
+    def load_projects(self) -> Iterator[dict]:
+        """Load all project configuration files from given directory
+
+        Yields:
+            Iterator[dict]:
+                Each element is a parsed YAML file as a dict
+        """
+        # Ignore all Sceptre resolvers
+        yaml.add_multi_constructor("!", lambda loader, suffix, node: None)
+        # Load the tower-project.yaml config files into a list
+        for config_path in self.list_projects():
+            with open(config_path) as config_file:
+                config = yaml.load(config_file, Loader=yaml.Loader)
+            if (
+                "template_path" in config
+                and config["template_path"] == "tower-project.yaml"
+            ):
+                yield config
+
+    def extract_emails(self, arns: Sequence[str]) -> List[str]:
+        """Extract role session names (emails) from assumed-role ARNs
+
+        Args:
+            arns (Sequence[str]): List of assumed-role ARNs
+
+        Returns:
+            List[str]: List of email from the role session names
+        """
+        role_arn_regex = re.compile(
+            r"arn:aws:sts::(?P<account_id>[0-9]+):assumed-role/(?P<role_name>[^/]+)"
+            r"/(?P<session_name>[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,})"
+        )
+        emails = list()
+        for arn in arns:
+            match = role_arn_regex.fullmatch(arn)
+            if match:
+                email = match.group("session_name")
+                emails.append(email)
+            else:
+                raise ValueError(
+                    f"Listed ARN ({arn}) doesn't follow expected format: "
+                    "'arn:aws:sts::<account_id>:<role_name>:<email>'"
+                )
+        return emails
+
+    def extract_users(self) -> Dict[str, Users]:
+        """Extract users from a series of config files
+
+        Returns:
+            Dict[str, Users]:
+                Mapping between projects/stacks and users
+        """
+        users_per_project = dict()
+        for config in self.load_projects():
+            assert "stack_name" in config
```

Comment:
Nit:  If you have a bunch of configurations and there is an invalid one in the middle, your code would fail for that one.  Are the templates already validated via precommit?

Reply:
What behavior is desirable? 

- Fail the entire script as soon as an invalid configuration is encountered
- Skip any invalid configurations, deploying the rest, and raise an exception at the end to note that some configs were invalid

We have some pre-commit hooks active in this repo, but is there a specific one that you have in mind?

Reply:
Hm... you could loop through the configuration to see if there is any invalid configuration first, then fail the script.  I guess I was thinking that there could be pre-commit hook with json schema validation, but that could be excessive here.

I leave it up to you to decide.  Your code does do a good job of not recreating anything that already exists, so I guess code failing due to invalid configurations doesn't really concern me.  Furthermore, if we are looking to only add one configuration at a time with PR's, the "only time" this code would fail would be if the new configuration added was invalid?

Reply:
I like this idea, but to avoid having this PR grow too big (and given that the risk of an invalid config is minimal, as you pointed out), I created a separate JIRA ticket for creating a JSON Schema for Tower project configs. 

https://sagebionetworks.jira.com/browse/WORKFLOWS-63

Reply:
If it helps, here's an example of yaml configuration validation I did for SC:
https://github.com/Sage-Bionetworks-IT/lambda-budgets/blob/master/budget/config.py

Reply:
Thanks! I added a link to this example code in the JIRA ticket.